File: apps/windows/tests/LexiconBar.Core.Tests/BurstDetectorTests.cs
using Xunit;

namespace LexiconBar.Tests;

/// <summary>
/// A port of <c>Tests/LexiconBarKitTests/BurstDetectorTests.swift</c>, case for
/// case. If a case here diverges from the Swift one, one of the two ports has
/// drifted and the two apps no longer behave the same way.
/// </summary>
public class BurstDetectorTests
{
    private const double Settle = 0.7;

    /// <summary>Silence that ends a streamed run (<c>Config.RunQuietMs</c>).</summary>
    private const double RunQuiet = 1.5;

    /// <summary>Types <paramref name="text"/> one character at a time, <paramref name="gap"/> seconds apart.</summary>
    private static double Type(string text, BurstDetector d, string baseText, double from, double gap = 0.08)
    {
        double now = from;
        string current = baseText;
        foreach (char ch in text)
        {
            current += ch;
            d.Record(current, now);
            now += gap;
        }
        return now;
    }

    private static Burst ExpectBurst(SettleOutcome outcome)
    {
        SettleOutcome.Fired fired = Assert.IsType<SettleOutcome.Fired>(outcome);
        return fired.Burst;
    }

    private static void ExpectSkipped(string reason, SettleOutcome outcome) =>
        Assert.Equal(new SettleOutcome.Skipped(reason), outcome);

    [Fact]
    public void CharAtATimeTypingNeverFires()
    {
        BurstDetector d = new(string.Empty);
        double end = Type("ping ashler about the cooper netties rollout", d, string.Empty, 0);
        ExpectSkipped("typed, not dictated", d.Settle(end + Settle));
        Assert.Equal("ping ashler about the cooper netties rollout", d.LastStable);
        // And a later poll finds nothing pending.
        Assert.Equal(SettleOutcome.Waiting.Instance, d.Settle(end + 5));
    }

    [Fact]
    public void FiveWordInsertionFiresAfterSettle()
    {
        BurstDetector d = new("Hi. ");
        d.Record("Hi. ping ashler about the rollout", 1.0);
        Assert.Equal(SettleOutcome.Waiting.Instance, d.Settle(1.3));
        Burst burst = ExpectBurst(d.Settle(1.0 + Settle));
        Assert.Equal("ping ashler about the rollout", burst.Text);
        Assert.Equal(new TextSpan(4, 29), burst.Span);
        Assert.Equal("Hi. ping ashler about the rollout", burst.FullText);
        Assert.Equal(SettleOutcome.Waiting.Instance, d.Settle(3));
    }

    [Fact]
    public void PhraseByPhraseDictationFires()
    {
        // Voice Access inserts words/phrases as they are recognised.
        BurstDetector d = new(string.Empty);
        d.Record("ping ", 0);
        d.Record("ping ashler ", 0.2);
        d.Record("ping ashler about the ", 0.4);
        d.Record("ping ashler about the cooper netties rollout", 0.6);
        Assert.Equal(SettleOutcome.Waiting.Instance, d.Settle(0.9));
        // A run that arrived as several insertions may still have words coming,
        // so SettleMs of quiet is no longer enough to end it.
        Assert.Equal(SettleOutcome.Coalescing.Instance, d.Settle(0.6 + Settle));
        Assert.Equal(string.Empty, d.LastStable);
        Burst burst = ExpectBurst(d.Settle(0.6 + RunQuiet));
        Assert.Equal(new TextSpan(0, 44), burst.Span);
    }

    [Fact]
    public void ShortSingleEventFiresAtTwelveChars()
    {
        BurstDetector d = new("Deploy to ");
        d.Record("Deploy to cooper netties", 0);
        Burst burst = ExpectBurst(d.Settle(Settle));
        Assert.Equal("cooper netties", burst.Text);

        BurstDetector shortRun = new(string.Empty);
        shortRun.Record("ashler", 0);
        // One word-sized insertion could be the first word of a dictated
        // phrase, so it is held until the silence says otherwise.
        Assert.Equal(SettleOutcome.Coalescing.Instance, shortRun.Settle(Settle));
        ExpectSkipped("too short (1 words, 6 units)", shortRun.Settle(RunQuiet));
    }

    [Fact]
    public void OwnRewriteDoesNotRetrigger()
    {
        BurstDetector d = new(string.Empty);
        d.Record("ping ashler about the rollout", 0);
        ExpectBurst(d.Settle(Settle));
        // The engine rewrites the field; the app sees a value change with the new text.
        d.MarkStable("ping Ashlr.AI about the rollout");
        d.Record("ping Ashlr.AI about the rollout", 1.0);
        Assert.False(d.HasPendingChanges);
        Assert.Equal(SettleOutcome.Waiting.Instance, d.Settle(1.0 + Settle));
        // Even without MarkStable, a value-changed event that lands on LastStable is inert.
        d.Record("ping Ashlr.AI about the rollout", 2.0);
        Assert.Equal(SettleOutcome.Waiting.Instance, d.Settle(3.0));
    }

    [Fact]
    public void MultiParagraphGuard()
    {
        BurstDetector d = new(string.Empty);
        d.Record("first line about ashler\nsecond line about versel", 0);
        ExpectSkipped("multi-paragraph", d.Settle(Settle));
        Assert.Equal("first line about ashler\nsecond line about versel", d.LastStable);

        BurstDetector trailing = new(string.Empty);
        trailing.Record("ping ashler about the rollout\n", 0);
        Burst b = ExpectBurst(trailing.Settle(Settle));
        Assert.Equal("ping ashler about the rollout", b.Text);
        Assert.Equal(new TextSpan(0, 29), b.Span);

        BurstDetector onlyNewline = new("abc");
        onlyNewline.Record("abc\n\n\n\n", 0);
        ExpectSkipped("nothing inserted", onlyNewline.Settle(Settle));

        Assert.True(BurstDetector.IsMultiParagraph("a\nb"));
        Assert.False(BurstDetector.IsMultiParagraph("a\n  \n"));
        // Windows line endings are the common case here, unlike on macOS.
        Assert.True(BurstDetector.IsMultiParagraph("a\r\nb"));
        Assert.False(BurstDetector.IsMultiParagraph("a\r\n"));
    }

    [Fact]
    public void CrLfDictationKeepsBothUnitsOutOfTheBurst()
    {
        // Notepad and most Win32 edits use CRLF. A dictated "new line" at the
        // end must leave the burst as a clean single paragraph, or the rewrite
        // would have to re-insert a line break it never owned.
        BurstDetector d = new(string.Empty);
        d.Record("ping ashler about the rollout\r\n", 0);
        Burst burst = ExpectBurst(d.Settle(Settle));
        Assert.Equal("ping ashler about the rollout", burst.Text);
        Assert.Equal(new TextSpan(0, 29), burst.Span);
    }

    [Fact]
    public void Utf16RangeMathWithEmoji()
    {
        string baseText = "Team \U0001F680 update: ";     // rocket is 2 UTF-16 units
        string inserted = "tell mason white about versel \U0001F44D";
        BurstDetector d = new(baseText);
        d.Record(baseText + inserted, 0);
        Burst burst = ExpectBurst(d.Settle(Settle));
        Assert.Equal(baseText.Length, burst.Span.Location);
        Assert.Equal(inserted.Length, burst.Span.Length);
        Assert.Equal(inserted, burst.Text);
        Assert.Equal(inserted, TextDiff.Substring(baseText + inserted, burst.Span));

        // Insertion in the middle, between two emoji.
        string mid = "\U0001F600\U0001F601";
        BurstDetector d2 = new(mid);
        d2.Record("\U0001F600ping ashler about the rollout\U0001F601", 0);
        Burst b2 = ExpectBurst(d2.Settle(Settle));
        Assert.Equal(new TextSpan(2, 29), b2.Span);
        Assert.Equal("ping ashler about the rollout", b2.Text);
    }

    [Fact]
    public void FieldTooLongIsSkipped()
    {
        BurstDetector d = new(string.Empty, new BurstDetector.Config(MaxFieldLength: 30));
        d.Record(string.Concat(Enumerable.Repeat("word ", 10)), 0);
        ExpectSkipped("field longer than 30", d.Settle(Settle));
    }

    [Fact]
    public void DeletionAndReplacementDoNotFire()
    {
        BurstDetector d = new("ping ashler about the rollout");
        d.Record("ping about the rollout", 0);
        ExpectSkipped("nothing inserted", d.Settle(Settle));
        // Selecting a word and dictating a single short word over it is too short.
        d.Record("ping Ashlr about the rollout", 1);
        Assert.Equal(SettleOutcome.Coalescing.Instance, d.Settle(1 + Settle));
        ExpectSkipped("too short (1 words, 6 units)", d.Settle(1 + RunQuiet));
    }

    [Fact]
    public void SettleHonoursConfiguredDelayAndMinWords()
    {
        BurstDetector d = new(string.Empty, new BurstDetector.Config(SettleMs: 1200, MinWords: 5, FewEventsMinChars: 100));
        d.Record("ping ashler about the", 0);
        Assert.Equal(SettleOutcome.Waiting.Instance, d.Settle(0.8));
        Assert.Equal(SettleOutcome.Coalescing.Instance, d.Settle(1.2));
        ExpectSkipped("too short (4 words, 21 units)", d.Settle(RunQuiet));
        d.Record("ping ashler about the cooper netties rollout", 2);
        // Diff against the new baseline: only " cooper netties rollout" is new (3 words), below MinWords 5.
        Assert.Equal(SettleOutcome.Coalescing.Instance, d.Settle(3.2));
        ExpectSkipped("too short (3 words, 23 units)", d.Settle(2 + RunQuiet));
    }

    // ---------------------------------------------------------------------
    // Streamed dictation
    //
    // Windows Voice Access, Wispr Flow and an app's own microphone button all
    // insert a word at a time rather than pasting a finished phrase, and the
    // gaps between spoken words are routinely longer than SettleMs. Deciding at
    // the first gap chopped such a phrase into single words, each "too short",
    // so the whole sentence went uncorrected.

    [Fact]
    public void StreamedDictationWithPausesLongerThanSettleCoalescesIntoOneBurst()
    {
        BurstDetector d = new(string.Empty);
        double now = 0.0;
        string text = string.Empty;
        foreach (string word in new[] { "ping ", "ashler ", "about ", "the ", "cuban ", "eats ", "rollout" })
        {
            text += word;
            d.Record(text, now);
            // The engine asks again SettleMs after each word; the run has to
            // stay pending or the word is dropped on its own.
            Assert.Equal(SettleOutcome.Coalescing.Instance, d.Settle(now + Settle, text.Length));
            now += 0.9;
        }

        double lastWordAt = now - 0.9;
        Burst burst = ExpectBurst(d.Settle(now + RunQuiet, text.Length));
        Assert.Equal("ping ashler about the cuban eats rollout", burst.Text);
        Assert.Equal(new TextSpan(0, 40), burst.Span);
        Assert.Equal(SettleOutcome.Waiting.Instance, d.Settle(lastWordAt + 5, text.Length));
    }

    [Fact]
    public void TypingThenPauseThenDictationCorrectsOnlyTheDictation()
    {
        BurstDetector d = new(string.Empty);
        // Typed by hand, character at a time: still refused.
        double now = Type("Meeting notes. ", d, string.Empty, 0);
        ExpectSkipped("typed, not dictated", d.Settle(now + Settle));
        Assert.Equal("Meeting notes. ", d.LastStable);

        // Then the user pauses and dictates. The burst is the dictated span
        // only — the typed prefix is already the baseline.
        string text = "Meeting notes. ";
        now += 2;
        foreach (string word in new[] { "ping ", "ashler ", "about ", "the ", "rollout" })
        {
            text += word;
            d.Record(text, now);
            Assert.Equal(SettleOutcome.Coalescing.Instance, d.Settle(now + Settle, text.Length));
            now += 0.9;
        }

        Burst burst = ExpectBurst(d.Settle(now + RunQuiet, text.Length));
        Assert.Equal("ping ashler about the rollout", burst.Text);
        Assert.Equal(new TextSpan(15, 29), burst.Span);
    }

    [Fact]
    public void SlowTypingIsNeverCoalescedIntoABurst()
    {
        // The guard that keeps the coalescing window from swallowing ordinary
        // typing is chunk size, not rate: typing one character at a time never
        // produces a MinChunk-sized insertion, however long the pauses are.
        BurstDetector d = new(string.Empty);
        double end = Type("ping ashler about the cuban eats rollout", d, string.Empty, 0, gap: 0.9);
        ExpectSkipped("typed, not dictated", d.Settle(end + Settle));
        Assert.Equal(SettleOutcome.Waiting.Instance, d.Settle(end + 10));
    }

    [Fact]
    public void APasteStillFiresAtSettleWithoutWaitingForTheRunWindow()
    {
        // One insertion that already reads as a burst has nothing to wait for;
        // holding it would add a second of latency to every paste.
        BurstDetector d = new(string.Empty);
        d.Record("ping ashler about the cuban eats rollout", 0);
        ExpectBurst(d.Settle(Settle));
    }

    [Fact]
    public void ACoalescingRunIsCappedByMaxRunMs()
    {
        // Continuous dictation must still be corrected periodically rather than
        // held forever waiting for a silence that never comes.
        BurstDetector d = new(string.Empty, new BurstDetector.Config(MaxRunMs: 3000));
        double now = 0.0;
        string text = string.Empty;
        foreach (string word in new[] { "ping ", "ashler ", "about ", "the ", "rollout " })
        {
            text += word;
            d.Record(text, now);
            now += 0.9;
        }

        ExpectBurst(d.Settle(now));
    }

    [Fact]
    public void HardRefusalsAreNotHeldOpen()
    {
        // Silence cannot turn a multi-paragraph insertion into a single one, so
        // it is refused at SettleMs rather than after the run window.
        BurstDetector d = new(string.Empty);
        d.Record("first line about ashler\nsecond line", 0);
        ExpectSkipped("multi-paragraph", d.Settle(Settle));
    }

    [Fact]
    public void TextDiffDelta()
    {
        Assert.Equal(new TextDiff.Delta(2, 2, 0), TextDiff.ComputeDelta("abc", "abXYc"));
        Assert.Equal(new TextDiff.Delta(3, 0, 0), TextDiff.ComputeDelta("abc", "abc"));
        Assert.Equal(new TextDiff.Delta(0, 3, 0), TextDiff.ComputeDelta(string.Empty, "abc"));
        // One more "a" could have been inserted at any of the four positions;
        // the window is reported with the play that makes it a guess.
        Assert.Equal(new TextDiff.Delta(3, 1, 0, SlideLeft: 3, SlideRight: 0), TextDiff.ComputeDelta("aaa", "aaaa"));
        Assert.Equal(new TextDiff.Delta(1, 1, 1), TextDiff.ComputeDelta("abc", "aXc"));
        Assert.Equal(2, TextDiff.WordCount("  two   words \n"));
    }
}

File: apps/macos/LexiconBar/Tests/LexiconBarKitTests/BurstDetectorTests.swift
import XCTest
@testable import LexiconBarKit

final class BurstDetectorTests: XCTestCase {
    private let settle = 0.7
    /// Silence that ends a streamed run (`Config.runQuietMs`).
    private let runQuiet = 1.5

    /// Types `text` one character at a time, `gap` seconds apart, starting at `t`.
    private func type(_ text: String, into d: BurstDetector, base: String, from t: TimeInterval, gap: TimeInterval = 0.08) -> TimeInterval {
        var now = t
        var current = base
        for ch in text {
            current.append(ch)
            d.record(text: current, at: now)
            now += gap
        }
        return now
    }

    func testCharAtATimeTypingNeverFires() {
        let d = BurstDetector(initialText: "")
        let end = type("ping ashler about the cooper netties rollout", into: d, base: "", from: 0)
        XCTAssertEqual(d.settle(at: end + settle), .skipped("typed, not dictated"))
        XCTAssertEqual(d.lastStable, "ping ashler about the cooper netties rollout")
        // And a later poll finds nothing pending.
        XCTAssertEqual(d.settle(at: end + 5), .waiting)
    }

    func testFiveWordInsertionFiresAfterSettle() {
        let d = BurstDetector(initialText: "Hi. ")
        d.record(text: "Hi. ping ashler about the rollout", at: 1.0)
        XCTAssertEqual(d.settle(at: 1.3), .waiting, "not quiet long enough")
        guard case .burst(let burst) = d.settle(at: 1.0 + settle) else { return XCTFail("expected a burst") }
        XCTAssertEqual(burst.text, "ping ashler about the rollout")
        XCTAssertEqual(burst.range, NSRange(location: 4, length: 29))
        XCTAssertEqual(burst.fullText, "Hi. ping ashler about the rollout")
        XCTAssertEqual(d.settle(at: 3), .waiting)
    }

    func testPhraseByPhraseDictationFires() {
        // macOS dictation inserts words/phrases as they are recognised.
        let d = BurstDetector(initialText: "")
        d.record(text: "ping ", at: 0)
        d.record(text: "ping ashler ", at: 0.2)
        d.record(text: "ping ashler about the ", at: 0.4)
        d.record(text: "ping ashler about the cooper netties rollout", at: 0.6)
        XCTAssertEqual(d.settle(at: 0.9), .waiting)
        // A run that arrived as several insertions may still have words
        // coming, so settleMs of quiet is no longer enough to end it.
        XCTAssertEqual(d.settle(at: 0.6 + settle), .coalescing)
        XCTAssertEqual(d.lastStable, "", "a coalescing run must not move the baseline")
        guard case .burst(let burst) = d.settle(at: 0.6 + runQuiet) else { return XCTFail("expected a burst") }
        XCTAssertEqual(burst.range, NSRange(location: 0, length: 44))
    }

    func testShortSingleEventFiresAtTwelveChars() {
        let d = BurstDetector(initialText: "Deploy to ")
        d.record(text: "Deploy to cooper netties", at: 0)
        guard case .burst(let burst) = d.settle(at: settle) else { return XCTFail("expected a burst") }
        XCTAssertEqual(burst.text, "cooper netties")

        let short = BurstDetector(initialText: "")
        short.record(text: "ashler", at: 0)
        // One word-sized insertion could be the first word of a dictated
        // phrase, so it is held until the silence says otherwise.
        XCTAssertEqual(short.settle(at: settle), .coalescing)
        XCTAssertEqual(short.settle(at: runQuiet), .skipped("too short (1 words, 6 units)"))
    }

    func testOwnRewriteDoesNotRetrigger() {
        let d = BurstDetector(initialText: "")
        d.record(text: "ping ashler about the rollout", at: 0)
        guard case .burst = d.settle(at: settle) else { return XCTFail("expected a burst") }
        // The engine rewrites the field; the app sees a value change with the new text.
        d.markStable("ping Ashlr.AI about the rollout")
        d.record(text: "ping Ashlr.AI about the rollout", at: 1.0)
        XCTAssertFalse(d.hasPendingChanges)
        XCTAssertEqual(d.settle(at: 1.0 + settle), .waiting)
        // Even without markStable, a value-changed event that lands on lastStable is inert.
        d.record(text: "ping Ashlr.AI about the rollout", at: 2.0)
        XCTAssertEqual(d.settle(at: 3.0), .waiting)
    }

    func testMultiParagraphGuard() {
        let d = BurstDetector(initialText: "")
        d.record(text: "first line about ashler\nsecond line about versel", at: 0)
        XCTAssertEqual(d.settle(at: settle), .skipped("multi-paragraph"))
        XCTAssertEqual(d.lastStable, "first line about ashler\nsecond line about versel")

        let trailing = BurstDetector(initialText: "")
        trailing.record(text: "ping ashler about the rollout\n", at: 0)
        guard case .burst(let b) = trailing.settle(at: settle) else { return XCTFail("a trailing newline is fine") }
        XCTAssertEqual(b.text, "ping ashler about the rollout", "the newline stays out of the burst")
        XCTAssertEqual(b.range, NSRange(location: 0, length: 29))
        let onlyNewline = BurstDetector(initialText: "abc")
        onlyNewline.record(text: "abc\n\n\n\n", at: 0)
        XCTAssertEqual(onlyNewline.settle(at: settle), .skipped("nothing inserted"))
        XCTAssertTrue(BurstDetector.isMultiParagraph("a\nb"))
        XCTAssertFalse(BurstDetector.isMultiParagraph("a\n  \n"))
    }

    func testUTF16RangeMathWithEmoji() {
        let base = "Team \u{1F680} update: "        // rocket is 2 UTF-16 units
        let inserted = "tell mason white about versel \u{1F44D}"
        let d = BurstDetector(initialText: base)
        d.record(text: base + inserted, at: 0)
        guard case .burst(let burst) = d.settle(at: settle) else { return XCTFail("expected a burst") }
        XCTAssertEqual(burst.range.location, base.utf16.count)
        XCTAssertEqual(burst.range.length, inserted.utf16.count)
        XCTAssertEqual(burst.text, inserted)
        XCTAssertEqual(TextDiff.substring(base + inserted, utf16: burst.range), inserted)
        // Insertion in the middle, between two emoji.
        let mid = "\u{1F600}\u{1F601}"
        let d2 = BurstDetector(initialText: mid)
        d2.record(text: "\u{1F600}ping ashler about the rollout\u{1F601}", at: 0)
        guard case .burst(let b2) = d2.settle(at: settle) else { return XCTFail("expected a burst") }
        XCTAssertEqual(b2.range, NSRange(location: 2, length: 29))
        XCTAssertEqual(b2.text, "ping ashler about the rollout")
    }

    func testFieldTooLongIsSkipped() {
        let d = BurstDetector(initialText: "", config: .init(maxFieldLength: 30))
        d.record(text: String(repeating: "word ", count: 10), at: 0)
        XCTAssertEqual(d.settle(at: settle), .skipped("field longer than 30"))
    }

    func testDeletionAndReplacementDoNotFire() {
        let d = BurstDetector(initialText: "ping ashler about the rollout")
        d.record(text: "ping about the rollout", at: 0)
        XCTAssertEqual(d.settle(at: settle), .skipped("nothing inserted"))
        // Selecting a word and dictating a single short word over it is too short.
        d.record(text: "ping Ashlr about the rollout", at: 1)
        XCTAssertEqual(d.settle(at: 1 + settle), .coalescing)
        XCTAssertEqual(d.settle(at: 1 + runQuiet), .skipped("too short (1 words, 6 units)"))
    }

    func testSettleHonoursConfiguredDelayAndMinWords() {
        let d = BurstDetector(initialText: "", config: .init(settleMs: 1200, minWords: 5, fewEventsMinChars: 100))
        d.record(text: "ping ashler about the", at: 0)
        XCTAssertEqual(d.settle(at: 0.8), .waiting)
        XCTAssertEqual(d.settle(at: 1.2), .coalescing)
        XCTAssertEqual(d.settle(at: runQuiet), .skipped("too short (4 words, 21 units)"))
        d.record(text: "ping ashler about the cooper netties rollout", at: 2)
        // Diff against the new baseline: only " cooper netties rollout" is new (3 words), below minWords 5.
        XCTAssertEqual(d.settle(at: 3.2), .coalescing)
        XCTAssertEqual(d.settle(at: 2 + runQuiet), .skipped("too short (3 words, 23 units)"))
    }

    // MARK: streamed dictation
    //
    // macOS dictation, Wispr Flow's streaming mode and an app's own microphone
    // button all insert a word at a time rather than pasting a finished
    // phrase, and the gaps between spoken words are routinely longer than
    // settleMs. Deciding at the first gap chopped such a phrase into single
    // words, each "too short", so the whole sentence went uncorrected.

    func testStreamedDictationWithPausesLongerThanSettleCoalescesIntoOneBurst() {
        let d = BurstDetector(initialText: "")
        var now = 0.0
        var text = ""
        for word in ["ping ", "ashler ", "about ", "the ", "cuban ", "eats ", "rollout"] {
            text += word
            d.record(text: text, at: now)
            // The engine asks again settleMs after each word; the run has to
            // stay pending or the word is dropped on its own.
            XCTAssertEqual(d.settle(at: now + settle, caretEnd: text.utf16.count), .coalescing,
                           "run ended early at \"\(word)\"")
            now += 0.9
        }
        let lastWordAt = now - 0.9
        guard case .burst(let burst) = d.settle(at: now + runQuiet, caretEnd: text.utf16.count) else {
            return XCTFail("expected the whole phrase as one burst")
        }
        XCTAssertEqual(burst.text, "ping ashler about the cuban eats rollout")
        XCTAssertEqual(burst.range, NSRange(location: 0, length: 40))
        XCTAssertEqual(d.settle(at: lastWordAt + 5, caretEnd: text.utf16.count), .waiting)
    }

    func testTypingThenPauseThenDictationCorrectsOnlyTheDictation() {
        let d = BurstDetector(initialText: "")
        // Typed by hand, character at a time: still refused.
        var now = type("Meeting notes. ", into: d, base: "", from: 0)
        XCTAssertEqual(d.settle(at: now + settle), .skipped("typed, not dictated"))
        XCTAssertEqual(d.lastStable, "Meeting notes. ")

        // Then the user pauses and dictates. The burst is the dictated span
        // only — the typed prefix is already the baseline.
        var text = "Meeting notes. "
        now += 2
        for word in ["ping ", "ashler ", "about ", "the ", "rollout"] {
            text += word
            d.record(text: text, at: now)
            XCTAssertEqual(d.settle(at: now + settle, caretEnd: text.utf16.count), .coalescing)
            now += 0.9
        }
        guard case .burst(let burst) = d.settle(at: now + runQuiet, caretEnd: text.utf16.count) else {
            return XCTFail("expected a burst")
        }
        XCTAssertEqual(burst.text, "ping ashler about the rollout")
        XCTAssertEqual(burst.range, NSRange(location: 15, length: 29))
    }

    func testSlowTypingIsNeverCoalescedIntoABurst() {
        // The guard that keeps the coalescing window from swallowing ordinary
        // typing is chunk size, not rate: typing one character at a time never
        // produces a minChunk-sized insertion, however long the pauses are.
        let d = BurstDetector(initialText: "")
        let end = type("ping ashler about the cuban eats rollout", into: d, base: "", from: 0, gap: 0.9)
        XCTAssertEqual(d.settle(at: end + settle), .skipped("typed, not dictated"))
        XCTAssertEqual(d.settle(at: end + 10), .waiting)
    }

    func testAPasteStillFiresAtSettleWithoutWaitingForTheRunWindow() {
        // One insertion that already reads as a burst has nothing to wait for;
        // holding it would add a second of latency to every paste.
        let d = BurstDetector(initialText: "")
        d.record(text: "ping ashler about the cuban eats rollout", at: 0)
        guard case .burst = d.settle(at: settle) else { return XCTFail("a paste must fire at settleMs") }
    }

    func testACoalescingRunIsCappedByMaxRunMs() {
        // Continuous dictation must still be corrected periodically rather
        // than held forever waiting for a silence that never comes.
        let d = BurstDetector(initialText: "", config: .init(maxRunMs: 3000))
        var now = 0.0
        var text = ""
        for word in ["ping ", "ashler ", "about ", "the ", "rollout "] {
            text += word
            d.record(text: text, at: now)
            now += 0.9
        }
        guard case .burst = d.settle(at: now) else { return XCTFail("expected the capped run to fire") }
    }

    func testHardRefusalsAreNotHeldOpen() {
        // Silence cannot turn a multi-paragraph insertion into a single one,
        // so it is refused at settleMs rather than after the run window.
        let d = BurstDetector(initialText: "")
        d.record(text: "first line about ashler\nsecond line", at: 0)
        XCTAssertEqual(d.settle(at: settle), .skipped("multi-paragraph"))
    }

    func testTextDiffDelta() {
        XCTAssertEqual(TextDiff.delta(from: "abc", to: "abXYc"), .init(location: 2, inserted: 2, removed: 0))
        XCTAssertEqual(TextDiff.delta(from: "abc", to: "abc"), .init(location: 3, inserted: 0, removed: 0))
        XCTAssertEqual(TextDiff.delta(from: "", to: "abc"), .init(location: 0, inserted: 3, removed: 0))
        // One more "a" could have been inserted at any of the four positions;
        // the window is reported with the play that makes it a guess.
        XCTAssertEqual(TextDiff.delta(from: "aaa", to: "aaaa"),
                       .init(location: 3, inserted: 1, removed: 0, slideLeft: 3, slideRight: 0))
        XCTAssertEqual(TextDiff.delta(from: "abc", to: "aXc"), .init(location: 1, inserted: 1, removed: 1))
        XCTAssertEqual(TextDiff.wordCount("  two   words \n"), 2)
    }
}
